server/models/w-user.js

```diff
@@ -0,0 +1,177 @@
+import { app } from '../server'
+import crypto from 'crypto'
+import * as ERRORS from '../lib/error'
+
+module.exports = WUser => {
+  WUser.disableRemoteMethodByName('find') // Disable GET /users
+  WUser.disableRemoteMethodByName('patchOrCreate') // Disable PATCH /users
+  WUser.disableRemoteMethodByName('replaceOrCreate') // Disable PUT /users
+  WUser.disableRemoteMethodByName('prototype.updateAttributes') // Disable PATCH /users/{id}
+  WUser.disableRemoteMethodByName('exists') // Disable HEAD /users/{id}
+  // Will Disable PUT /users/{id} at same time
+  // WUser.disableRemoteMethodByName('replaceById') // Disable POST /users/{id}/replace
+  WUser.disableRemoteMethodByName('createChangeStream') // Disable GET&POST /users/{id}/change-stream
+  WUser.disableRemoteMethodByName('count') // Disable GET /users/count
+  WUser.disableRemoteMethodByName('findOne') // Disable GET /users/findOne
+  WUser.disableRemoteMethodByName('updateAll') // Disable POST /users/update
+  WUser.disableRemoteMethodByName('upsertWithWhere') // Disable POST /users/upsertWithWhere
+
+  const makeSalt = () => Math.round((new Date().valueOf() * Math.random())) + ''
+  const encryptPassword = (password, salt) =>
+    crypto.createHmac('sha1', salt)
+      .update(password)
+      .digest('hex')
+
+  WUser.prototype.hasPassword = function (plain) {
+    if (this.password && plain) {
+      return encryptPassword(plain, this.salt) === this.password
+    }
+  }
+
+  WUser.prototype.createToken = function (ttl, cb) {
+    if (typeof ttl === 'function') {
+      cb = ttl
+      ttl = undefined
+    }
+    const { WToken } = app.models
+
+    WToken.create({
+      ttl: ttl,
+      userId: this.id
+    }, (err, instance) => {
+      cb(err, instance)
+    })
+  }
+
+  WUser.observe('before save', (ctx, next) => {
+    const { WUser } = app.models
+    if (ctx.instance && ctx.instance.user_name && ctx.instance.name) {
+      if (ctx.isNewInstance) {
+        // Create new user
+        WUser.findOne({
+          where: {
+            user_name: ctx.instance.user_name
+          }
+        }, (err, instance) => {
+          if (err) return next(ERRORS.MONGO_ERROR(err))
+
+          if (instance) {
+            // Duplicated user name
+            next(ERRORS.USER_NAME_BEEN_TAKEN)
```

Comment:
For the TOCTOU problem i think it is more like a competition problem that happens in concurrent programming, but with DB, i think it should be the DB's responsibility to take over these problems, like the atomic method given in MongoDB.

But to avoid this problem in code logic, i think the following solutions
 * the way that solve this problem in concurrent programming can be refer,build a lock system to make sure the item is safe before any operation taken, and unlocked after operation success, but this also need to consider how to make the lock system work fast.

* A more easier way to avoid this kind of problem, is to use a queue if these logic can take the waiting time in the queue. a queue is more secure way deal with these kind of problem, all the operations are happens synchronously, there will be no worries about the competition.